idea/src/org/jetbrains/kotlin/idea/editor/KotlinEnterHandler.kt

```diff
@@ -122,3 +124,47 @@ class KotlinEnterHandler: EnterHandlerDelegateAdapter() {
         return false
     }
 }
+
+class KotlinConvertToBodyEnterHandler : EnterHandlerDelegateAdapter() {
+    override fun preprocessEnter(file: PsiFile, editor: Editor,
+                                 caretOffset: Ref<Int>, caretAdvance: Ref<Int>, dataContext: DataContext,
+                                 originalHandler: EditorActionHandler?): EnterHandlerDelegate.Result {
+        if (file !is KtFile) return EnterHandlerDelegate.Result.Continue
+        if (!KotlinEditorOptions.getInstance().isEnableSmartConvertToBody) return EnterHandlerDelegate.Result.Continue
+        if (!EnterAfterUnmatchedBraceHandler.isAfterUnmatchedLBrace(editor, caretOffset.get(),
+                                                                    file.fileType)) return EnterHandlerDelegate.Result.Continue
+
+        val document = editor.document
+
+        //Let's check if it's special convert to body case.
+        val prevOffset = caretOffset.get() - 1
+        if (document.isTextAt(prevOffset, "{")) {
+            document.commit(file.project)
+            val element = file.findElementAt(prevOffset)
+            if (element != null && element.matchParents(
+                    KtFunctionLiteral::class.java,
+                    KtLambdaExpression::class.java,
+                    KtDeclarationWithBody::class.java
+            )) {
+                val declaration = element.parent.parent.parent as KtDeclarationWithBody
+                if (!declaration.hasDeclaredReturnType()) return EnterHandlerDelegate.Result.Continue
+                val equals = declaration.equalsToken
+                if (equals != null) {
+                    val startOffset =
```

Comment:
Will it work with comments? Maybe it would be better to check for '=' and extend it with spaces from left and right.